src/transform/src/canonicalize_mfp.rs

```diff
@@ -272,37 +276,69 @@ impl CanonicalizeMfp {
         if or_args.len() == 0 {
             return false;
         }
-        let or_args_residual = or_args
-            .iter()
-            .map(|or_arg| {
-                let mut and_args = or_arg.and_or_args(VariadicFunc::And);
-                // Note that `remove_impossible_or_args` made sure that inside each or_arg, each
-                // expression can have only one literal constraint. So if we find one of the
-                // key fields being literal constrained, then it's definitely that literal
-                // constraint that detect_literal_constraints based it's return values on.
-                //
-                // This is important, because without `remove_impossible_or_args`, we might
-                // have the situation here that or_arg would be something like
-                // `a = 5 AND a = 8`, of which `detect_literal_constraints` found only the `a = 5`,
-                // but here we would remove both the `a = 5` and the `a = 8`.
-                and_args.retain(|and_arg| {
-                    !key.iter()
+
+        // In simple situations it would be enough to check here that if we remove the detected
+        // literal constraints from each OR arg, then the residual OR args are all equal.
+        // However, this wouldn't be able to perform the removal when the expression that should
+        // remain in the end has an OR. This is because conversion to DNF makes duplicates of
+        // every literal constraint, with different residuals. To also handle this case, we collect
+        // the possible residuals for every literal constraint row, and check that all sets are
+        // equal. Example: The user wrote
+        // `WHERE ((a=1 AND b=1) OR (a=2 AND b=2)) AND (c OR (d AND e))`.
+        // The DNF of this is
+        // `(a=1 AND b=1 AND c) OR (a=1 AND b=1 AND d AND e) OR (a=2 AND b=2 AND c) OR (a=2 AND b=2 AND d AND e)`.
+        // Then `constraints_to_residual_sets` will be:
+        // [
+        //   [`a=1`, `b=1`]  ->  {[`c`], [`d`, `e`]},
+        //   [`a=2`, `b=2`]  ->  {[`c`], [`d`, `e`]}
+        // ]
+        // After removing the literal constraints we have
+        // `c OR (d AND e)`
+        let mut constraints_to_residual_sets = BTreeMap::new();
+        or_args.iter().for_each(|or_arg| {
+            let and_args = or_arg.and_or_args(VariadicFunc::And);
+            let (mut constraints, mut residual): (Vec<_>, Vec<_>) =
+                and_args.iter().partition(|and_arg| {
```

Comment:
I think if you change this line to `and_args.iter().cloned().partition(|and_arg| {`, then you won't need the `clone_elements` calls later or the `clone_elements` function.

Reply:
Done